catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrMetacardClientImpl.java

```diff
@@ -296,11 +287,36 @@ protected SolrQuery postAdapt(QueryRequest request, SolrFilterDelegate filterDel
         query.setStart(request.getQuery()
                 .getStartIndex() - 1);
 
+        if (queryingForAllRecords(request)) {
+            try {
+                int numRows = queryForNumberOfRows(query);
+                query.setRows(numRows);
+            } catch (SolrServerException | IOException | SolrException exception) {
+                throw new UnsupportedQueryException("Could not complete solr query.", exception);
+            }
+        } else {
+            query.setRows(request.getQuery()
+                    .getPageSize());
+        }
+
         setSortProperty(request, query, filterDelegate);
 
         return query;
     }
 
+    private boolean queryingForAllRecords(QueryRequest request) {
+        return request.getQuery()
+                .getPageSize() < 1;
+    }
+
+    private int queryForNumberOfRows(SolrQuery query) throws SolrServerException, IOException {
+        query.setRows(0);
+        QueryResponse solrResponse = client.query(query, SolrRequest.METHOD.POST);
+        int numRows = (int) solrResponse.getResults()
```

Comment:
Instead of the cast, would it be safer to do something like this? http://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#toIntExact-long-